core/deployment/src/main/java/io/quarkus/deployment/steps/DevServicesConfigBuildStep.java

```diff
@@ -60,10 +62,21 @@ public void run() {
                 }
             }, true);
         }
-        for (Map.Entry<String, String> entry : newProperties.entrySet()) {
+        oldConfig = newProperties;
+
+        Map<String, String> newPropertiesWithAdditionalConfig = new HashMap<>(newProperties);
+        // On contrary to dev services config, "additional" config build items are
+        // produced on each restart, so we don't want to remember them from one restart to the next.
+        for (DevServicesAdditionalConfigBuildItem item : devServicesAdditionalConfigBuildItems) {
+            if (newProperties.containsKey(item.getTriggeringKey())) {
+                item.getCallbackWhenEnabled().run();
```

Comment:
Potential NPE here

Reply:
Not at the moment, but I guess you never know where people will try to stuff a `null`... I'll add a check.